pageserver/src/tenant.rs

```diff
@@ -2569,12 +2570,13 @@ impl Tenant {
         horizon: u64,
         pitr: Duration,
         ctx: &RequestContext,
+        cancel: &CancellationToken,
```

Comment:
This function does a fair amount of work iterating through each timeline, without checking for cancellation. The cancellation token is only passed to refresh_gc_info_internal(), and not checked after that.

That's not wrong as such, but it means that the function might run a lot longer than you'd think, after cancellation has been requested. The loop does check for `task_mgr::is_shutdown_requested()`, so that would probably be a good place to check for the cancellation token too.

Reply:
37895c21fa8e663b91ba594170654f168fa081c9